### ISSUE TITLE
--rm flag in docker run commands

### ISSUE THREAD

**Issue by lebannen** [jupyter/docker-stacks#476]
**What docker image you are using?**

`jupyter/tensorflow-notebook`

**What complete docker command do you run to launch the container (omitting sensitive values)?**

Example: `docker run -it --rm -p 8888:8888 jupyter/tensorflow-notebook`

**What do you expect to happen?**

I didn't expect --rm flag in basic use command. I've started docker after restart and have not found my container. Then I've realized that command has --rm flag, so this behaviour is obviously correct, but it's not very pleasant to lose all your work after restart.

**What actually happens?**
Container is removed after restart.

Is it possible to remove --rm flag from at least basic use commands?

**Comment by charlesreid1** [comment]
You can map a local directory to the Docker container's `/home/jovyan` directory using the `-v` flag, which will prevent losing your work. Here's an example script I use, which I drop into a Github repository. The `notebooks/` directory now allows version-controlled Jupyter notebooks to be passed through.

```
#!/bin/bash
# Run the jupyter/tensorflow-notebook
docker run -it --rm \
    -v ${PWD}/notebooks:/home/jovyan \
    -p 8888:8888 jupyter/tensorflow-notebook
```

**Comment by jakirkham** [comment]
Indeed. @charlesreid1 is correct. Mapping directories from the native machine to the Docker container is the way to go. Any other questions @lebannen or can we close this?

**Comment by lebannen** [comment]
@jakirkham my problem was that I've lost results of one day of my work because of this flag. I understand that it was my fault and I should be more careful with docker command. I know how to handle that, I've just removed -it --rm, because I don't need them.
My idea is that when you have default command in readme, it should just work. Without mounting volumes or doing other docker staff.

**Comment by parente** [comment]
We can switch to not specifying the `--rm` in the default command. If we do, we'll have to add a note telling users that if they continually create containers without `--rm` and don't use `docker rm <container name or id>` or `docker system prune` routinely, they'll silently fill disk with stopped containers.

**Comment by lebannen** [comment]
@parente Why should anyone create container each time? I've created container once and it keeps working. If I need to stop/start it, I use `docker stop <container name>`/`docker start <container name>`.

**Comment by charlesreid1** [comment]
I think the `--rm` flag should remain, as it is in keeping with the principle of [disposability](https://12factor.net/disposability) (not to mention, `--rm` is a flag included by the Docker developers wherever they start containers in the documentation). Users shouldn't rely on the container maintaining a persistent state and sticking around indefinitely. That being said, perhaps a warning or some advice on how to persist data (like the `-v` flag) could be added.

**Comment by lebannen** [comment]
@charlesreid1 
1) how it provides

> fast startup and graceful shutdown

It just destroys container with all its state, which makes startup slower in comparison with `docker start` after `docker stop` without `--rm` flag. If you check [this docs](https://github.com/docker/labs/blob/master/12factor/09_disposability.md) there is not a single word about `--rm` flag.

2) Where do Docker developers add this flag? If you check [official documentation](https://docs.docker.com/engine/reference/run/#clean-up-rm), then you will see 

> But if you are running short-term foreground processes, these container file systems can really pile up. If instead you’d like Docker to automatically clean up the container and remove the file system when the container exits, you can add the --rm flag

3) Why shouldn't user rely on the container state? When you write something in notebook in real life, you don't throw it in a recycle bin and buy new notebook. I expect similar behavior here as well. There are a lot of images for databases in docker hub and they don't have `--rm` flag in basic commands ([example1](https://hub.docker.com/_/mysql/), [example2](https://hub.docker.com/_/postgres/)). They have `--rm` flag for clients, because you don't want to keep after client exits, but not for the db itself.

**Comment by charlesreid1** [comment]
As the docker documentation and @parente both mention, leaving off the `--rm` flag leaves resources laying around when containers are created/destroyed. In a production system with many containers being created/destroyed this *does* affect startup time. 

From your description it sounds like your use case is a single user using the container on a single machine, in which case your arguments are perfectly valid; they are not valid when extended to a production system with many containers being created/destroyed. Those users and use cases are why I'm arguing for the `--rm` flag being left on. The `-v` flag will solve your problem, so you aren't left without resolution.

I understand you had a frustrating experience with your container being deleted and your work lost. But that does not justify removal of the `--rm` flag from the documentation, which would affect the use of docker-stacks and the resources they consume in larger scale production environments. The proper resolution is the use of the `-v` flag.

Docker itself is designed not to be stateful - that's the purpose of containers. Virtual machines are ephemeral so that you don't have to deal with a stateful machine. Otherwise, why not use a fleet of virtual servers? Docker provides mechanisms like `-v` and [Docker volumes](https://docs.docker.com/engine/admin/volumes/volumes/) to get around the statelessness of containers, but statelessness is the way of container-zen. Containers holding on to resources even after they are destroyed is not.

**Comment by lebannen** [comment]
@charlesreid1 as I mentioned before, in production system you will **never** run containers with `docker run` multiple times. You create containers once and then start/stop them. 
You use volumes in production systems, because docker file system (why it has its own file system if it was designed stateless?) is much less efficient with a lot of changes. It does not mean that you can easily use `--rm`. Container may have some inner logic to restore some state from existing data.

In any case these notebooks are not designed to be used with other containers and multiple users. Or at least it's not their basic use case.
